contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,944 @@
+    ISwapRouter public immutable swapRouter;
+    /// @notice the underlying AMO Slipstream pool
+    ICLPool public immutable clPool;
+    /// @notice the gauge for the corresponding Slipstream pool (clPool)
+    /// @dev can become an immutable once the gauge is created on the base main-net
+    ICLGauge public immutable clGauge;
+    /// @notice the Position manager contract that is used to manage the pool's position
+    INonfungiblePositionManager public immutable positionManager;
+    /// @notice helper contract for liquidity and ticker math
+    ISugarHelper public immutable helper;
+    /// @notice sqrtRatioX96TickLower
+    /// @dev tick lower has value -1 and represents the lowest price of WETH priced in OETHb. Meaning the pool
+    /// offers less than 1 OETHb for 1 WETH. In other terms to get 1 OETHB the swap needs to offer 1.0001 WETH
+    /// this is where purchasing OETHb with WETH within the liquidity position is most expensive
+    uint160 public immutable sqrtRatioX96TickLower;
+    /// @notice sqrtRatioX96TickHigher
+    /// @dev tick higher has value 0 and represents 1:1 price parity of WETH to OETHb
+    uint160 public immutable sqrtRatioX96TickHigher;
+    /// @dev tick closest to 1:1 price parity
+    ///      Correctly assesing which tick is closer to 1:1 price parity is important since it affects
+    ///      the way we calculate the underlying assets in check Balance
+    uint160 public immutable sqrtRatioX96TickClosestToParity;
+
+    error NotEnoughWethForSwap(uint256 wethBalance, uint256 requiredWeth); // 0x989e5ca8
+    error NotEnoughWethLiquidity(uint256 wethBalance, uint256 requiredWeth); // 0xa6737d87
+    error PoolRebalanceOutOfBounds(
+        uint256 currentPoolWethShare,
+        uint256 requiredPoolWethShare
+    ); // 0x6c6108fb
+    error OutsideExpectedTickRange(int24 currentTick); // 0x46a58db6
+
+    event PoolRebalanced(
+        uint256 currentPoolWethShare,
+        uint256 targetPoolWethShare
+    );
+
+    event PoolWethShareUpdated(uint256 newWethShare);
+
+    event PrincipalPositionAfterSwap(
+        uint256 wethPositionBalance,
+        uint256 oethbPositionBalance
+    );
+
+    event PoolWethShareVarianceAllowedUpdated(
+        uint256 poolWethShareVarianceAllowed
+    );
+
+    event LiquidityRemoved(
+        uint256 withdrawLiquidityShare,
+        uint256 removedWETHAmount,
+        uint256 removedOETHbAmount,
+        uint256 wethAmountCollected,
+        uint256 oethbAmountCollected,
+        uint256 underlyingAssets
+    );
+
+    event LiquidityAdded(
+        uint256 wethAmountDesired,
+        uint256 oethbAmountDesired,
+        uint256 wethAmountSupplied,
+        uint256 oethbAmountSupplied,
+        uint256 tokenId,
+        uint256 underlyingAssets
+    );
+
+    event UnderlyingAssetsUpdated(uint256 underlyingAssets);
+
+    /**
+     * @dev Verifies that the caller is the Governor, or Strategist.
+     */
+    modifier onlyGovernorOrStrategist() {
```

Comment:
⚪️ When we use this modifier, we use it for strategist actions with governance backup. 99% of the time it's the strategist calling it. This if we have the strategist check first, we save 2,000 gas on every admin action. See the vaultAdmin code:

https://github.com/OriginProtocol/origin-dollar/blob/8b3c95016f734797894ee7a3a1d020aff169874a/contracts/contracts/vault/VaultAdmin.sol#L26-L32

Reply:
Nice touch to make use of the gas efficient [short curcuit](https://medium.com/@t.tak/how-to-reduce-gas-cost-in-solidity-f2e5321e0395#9dac) properties. Done here: https://github.com/OriginProtocol/origin-dollar/pull/2184/commits/a9e4f8b26987ceeaed3cc36b26336df17e9c2ae6